lib/SILGen/SILGenPattern.cpp

```diff
@@ -2555,19 +2557,77 @@ void SILGenFunction::usingImplicitVariablesForPattern(Pattern *pattern, CaseStmt
   variableSwapper();
 }
 
+static void emitDiagnoseOfUnexpectedEnumCaseValue(SILGenFunction &SGF,
+                                                  SILLocation loc,
+                                                  ManagedValue value,
+                                                  const EnumDecl *enumDecl) {
+  ASTContext &ctx = SGF.getASTContext();
+  auto diagnoseFailure = ctx.getDiagnoseUnexpectedEnumCaseValue(nullptr);
+  if (!diagnoseFailure) {
+    SGF.B.createBuiltinTrap(loc);
+    return;
+  }
+
+  assert(enumDecl->isObjC());
+  assert(enumDecl->hasRawType());
+
+  // Get the enum type as an Any.Type value.
+  SILType metatypeType = SILType::getPrimitiveObjectType(
+      CanMetatypeType::get(value.getType().getSwiftRValueType(),
+                           MetatypeRepresentation::Thick));
+  SILValue metatype = SGF.B.createMetatype(loc, metatypeType);
+  SILType anyTypeType = SGF.getLoweredType(
+      ExistentialMetatypeType::get(ctx.TheAnyType));
+  SILValue upcastMetatype =
+      SGF.B.createInitExistentialMetatype(loc, metatype, anyTypeType, None);
+
+  // Bitcast the enum value to its raw type. (This is only safe for @objc
+  // enums.)
```

Comment:
Do we ever end up here for stringly-typed enums? Maybe assert that the raw type is trivial

Reply:
Ah, no, we don't. Those remain structs. But I'll add the assertion.